File: 0994-rotting-oranges/0994-rotting-oranges.py
from collections import deque
from typing import List


class Solution:

    def orangesRotting(self, grid: List[List[int]]) -> int:
        queue = deque()
        rows = len(grid)
        cols = len(grid[0])
        fresh_oranges = 0
        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == 2:
                    queue.append((r, c))
                elif grid[r][c] == 1:
                    fresh_oranges += 1

        queue.append((-1, -1))
        minutes_elapsed = -1
        directions = [(1, 0), (-1, 0), (0, -1), (0, 1)]

        while queue:
            row, col = queue.popleft()
            if row == -1:
                minutes_elapsed += 1
                if queue:
                    queue.append((-1, -1))
            else:
                for d in directions:
                    neighbor_row = row + d[0]
                    neighbor_col = col + d[1]
                    if rows > neighbor_row >= 0 and cols > neighbor_col >= 0:
                        if grid[neighbor_row][neighbor_col] == 1:
                            grid[neighbor_row][neighbor_col] = 2
                            fresh_oranges -= 1
                            queue.append((neighbor_row, neighbor_col))

        return minutes_elapsed if fresh_oranges == 0 else -1

File: 0994-rotting-oranges/0994-rotting-oranges.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
var orangesRotting = function(grid) {
    let queue = [];
    let rows = grid.length;
    let cols = grid[0].length;
    let freshOranges = 0;

    // Populate queue with initial rotten oranges and count fresh oranges
    for (let r = 0; r < rows; r++) {
        for (let c = 0; c < cols; c++) {
            if (grid[r][c] === 2) {
                queue.push([r, c]); // Add rotten orange to the queue
            } else if (grid[r][c] === 1) {
                freshOranges += 1; // Count fresh oranges
            }
        }
    }

    queue.push([-1, -1]); // Marker to indicate minute elapsed
    let minutesElapsed = -1;
    const directions = [[1, 0], [-1, 0], [0, -1], [0, 1]]; // Direction vectors for adjacent cells

    // BFS traversal using the queue
    while (queue.length > 0) {
        let [row, col] = queue.shift();
        if (row === -1) {
            // Increment time when marker is encountered
            minutesElapsed += 1;
            if (queue.length > 0) {
                queue.push([-1, -1]); // Add marker back if there are still oranges to process
            }
        } else {
            // Process adjacent cells
            for (let d of directions) {
                let nextRow = row + d[0];
                let nextCol = col + d[1];
                // Check bounds and if adjacent cell contains a fresh orange
                if (nextRow >= 0 && nextRow < rows && nextCol >= 0 && nextCol < cols && grid[nextRow][nextCol] === 1) {
                    grid[nextRow][nextCol] = 2; // Mark orange as rotten
                    freshOranges -= 1; // Decrease fresh orange count
                    queue.push([nextRow, nextCol]); // Add newly rotten orange to the queue
                }
            }
        }
    }

    // If there are still fresh oranges, return -1, otherwise return the time elapsed
    return freshOranges === 0 ? minutesElapsed : -1;
};

// Example usage
let grid = [[2,1,1],[1,1,0],[0,1,1]];
console.log(orangesRotting(grid)); // Output: 4
